src/library/coverart.h

```diff
@@ -30,48 +31,80 @@ struct CoverInfo {
 
     static const quint16 kNullImageHash;
 
-    CoverInfo() : source(UNKNOWN),
-                  type(NONE),
-                  coverLocation(QString()),
-                  trackLocation(QString()),
-                  // This default value is fine: qChecksum(NULL, 0) is 0.
-                  hash(CoverInfo::kNullImageHash) {}
+    CoverInfoRelative()
+            : source(UNKNOWN),
+              type(NONE),
+              // This default value is fine: qChecksum(NULL, 0) is 0.
+              hash(kNullImageHash) {
+    }
 
-    bool operator==(const CoverInfo& other) const {
+    bool operator==(const CoverInfoRelative& other) const {
         return other.source == source &&
                 other.type == type &&
                 other.coverLocation == coverLocation &&
-                other.trackLocation == trackLocation &&
                 other.hash == hash;
     }
-    bool operator!=(const CoverInfo& other) const {
+    bool operator!=(const CoverInfoRelative& other) const {
         return !(*this == other);
     }
 
     Source source;
     Type type;
-    QString coverLocation; // Relative path, starting from trackLocation
-    QString trackLocation;
+    QString coverLocation; // relative path, from track location
     quint16 hash;
 };
 
-struct CoverArt {
-    CoverArt() {}
+class CoverInfo : public CoverInfoRelative {
+  public:
+    CoverInfo() {}
+
+    CoverInfo(const CoverInfoRelative& base, const QString& tl)
```

Comment:
I don't insist on eliminating the class hierarchy ;) It was just an idea to 
think about.

It would be nice to leverage the inheritance hierarchy by encapsulating 
varying behavior and differences in handling directly in polymorphic virtual 
functions. I'm still not convinced of the chosen class names. They should 
reveal the abstract concept behind each class, and not any implementation 
details.

On 17.08.2016 23:37, Daniel Schürmann wrote:
>
> In src/library/coverart.h 
> <https://github.com/mixxxdj/mixxx/pull/990#discussion_r75212327>:
>
> >      quint16 hash;
> >  };
> >
> > -struct CoverArt {
> > -    CoverArt() {}
> > +class CoverInfo : public CoverInfoRelative {
> > +  public:
> > +    CoverInfo() {}
> > +
> > +    CoverInfo(const CoverInfoRelative& base, const QString& tl)
>
> Yes, you are right, it is composing data.
>
> I have squeezed this forth and back. The problem in this case with a 
> composition is that three classes are involved. The member access will be 
> either cluttered like this:
> coverArt.CoberInfo.CoverInforelative.hash or we would need a bunch of 
> redundant getter and setter functions, which hide the composition nature 
> anyway.
> A composition becomes handy, if the composed member is a base class of a 
> set of possible classes, Since this is not the case here, I like to keep 
> the Inheritance for now.
> I have added the missing virtual destructors.
>
> Does this work for you? If not, I will go the extra mile and introduce the 
> getter and setter for the composing model.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/990/files/51afd202825acec141e5fe9a621015f3bae0475f#r75212327>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABZ-aqoczlufQEvoCd1BI3JIQ6tXaggOks5qg38ggaJpZM4JeH62>.
>